### PR TITLE
Overhaul the `RWST` monad and add support for reasoning with Dijkstra's weakest precondition calculus.

### PR DESCRIPTION
**Pull request by cwjnkins**
This PR also:

- adds a `Monad` record opened with Agda's instance arguments, to simulate Haskell's type classes (this allows us to use do-notation for `Either` as well as `LBFT`)
- moves the convenient syntax for do-notation with `RWST` to its own file. This module is imported publicly by `LibraBFT/ImplShared/Util/Util.agda`, so many occurrences of `open RWST-do` have been removed from the implementation models.
- replaces some `if` and `case` statements with the branching operators of the new `RWST` datatype
- removes `ErrLog` and replaces all remaining occurrences with `FakeErr`

### CODE REVIEW

File: LibraBFT/ImplShared/Util/RWST.agda
Comment (mark-moir) [inline LibraBFT/ImplShared/Util/RWST.agda:14]: s/aide/aid/

File: LibraBFT/ImplShared/Util/RWST.agda
Comment (mark-moir) [inline LibraBFT/ImplShared/Util/RWST.agda:59]: s/prove/provide/ ?

File: LibraBFT/ImplShared/Util/RWST.agda
Comment (mark-moir) [inline LibraBFT/ImplShared/Util/RWST.agda:160]: s/provide/prove/ ?

File: LibraBFT/ImplShared/Util/RWST.agda
Comment (mark-moir) [inline LibraBFT/ImplShared/Util/RWST.agda:168]: It would be helpful to add something like: This proves that our definition of `RWST-weakestPre` provides a _sufficient_ precondition for the post condition to hold.  It does _not_ prove that it is the weakest such precondition (even though it's true); this is not important for our purposes.

**Review by mark-moir — APPROVED** [review]
Looks great @cwjnkins, thanks!  I've just made a few minor comments on spelling and comments.  Feel free to address as you see fit and merge.

**Comment by cwjnkins** [comment]
@mark-moir I have made the requested changes. I will wait until later today (after we meet) to merge this in.

File: LibraBFT/Impl/Consensus/RoundManager.agda
Comment (haroldcarr) [inline LibraBFT/Impl/Consensus/RoundManager.agda:57]: What does `MM` mean?
Why isn't `case` good enough here?

File: LibraBFT/Impl/Consensus/RoundManager.agda
Comment (haroldcarr) [inline LibraBFT/Impl/Consensus/RoundManager.agda:89]: I'm definitely confused.

In the haskell code, the `M` suffix was generally used to mean functions in the `LBFT` monad.

This particular `ifM` seems that it is turning the decidable test into a boolean.   That has nothing to do with a monad nor a `LBFT` monad.

File: LibraBFT/Impl/Consensus/RoundManager.agda
Comment (haroldcarr) [inline LibraBFT/Impl/Consensus/RoundManager.agda:126]: If I am right about my earlier comment, then this should be `if||`

**Comment by haroldcarr** [comment]
before I continue reviewing, I am going to wait for responses to my existing comments since I seem to be missing some context

**Comment by haroldcarr** [comment]
Approved - pending answer to why `FakeErr` rather than `ErrLog`

**Review by haroldcarr — APPROVED** [review]

File: LibraBFT/Impl/Consensus/BlockStorage/SyncManager.agda
Comment (cwjnkins) [reply LibraBFT/Impl/Consensus/BlockStorage/SyncManager.agda:23]: This was the name @mark-moir introduced. I am also not sure why it is `FakeErr`.

Comment (mark-moir) [reply LibraBFT/Impl/Consensus/BlockStorage/SyncManager.agda:23]: I'm sure we discussed this, but for the record...

The point of `FakeErr` is to serve as a placeholder in our model (Agda implementation of the Haskell code) for where the Haskell code logs actual errors, which have a bunch of constructors, most of which take `[Text]` arguments.  We don't want to spend time modeling these errors in detail at this stage (or maybe ever, at least in cases we can prove don't ever happen, e.g., internal invariant checks).  We were initially not modeling the errors at all, but this would affect the "shape" of proofs, so "we" decided to have a placeholder with no structure.  HTH.

Comment (mark-moir) [reply LibraBFT/Impl/Consensus/BlockStorage/SyncManager.agda:23]: Perhaps a better way would be to keep `ErrLog`, but just model it with a single constructor `FakeErr`.  That way we could incrementally add real constructors later, perhaps eventually eliminating the`FakeErr` constructor, or spell it differently such as `ImpossibleError`, and then prove that nothing ever emits `ImpossibleError`.  Thoughts?

Comment (cwjnkins) [reply LibraBFT/Impl/Consensus/BlockStorage/SyncManager.agda:23]: This was the name @mark-moir introduced. I am also not sure why it is `FakeErr`.

Comment (cwjnkins) [reply LibraBFT/Impl/Consensus/BlockStorage/SyncManager.agda:23]: I don't think there is a need to prove that `ImpossibleError` is never emitted when all the errors have been implemented -- we can just remove that constructor. I think having a datatype for `ErrLog` and *postulating* `fakeErr` (so that no proofs ever match against it) might be a good way to go.

Comment (mark-moir) [reply LibraBFT/Impl/Consensus/BlockStorage/SyncManager.agda:23]: My point was that maybe there's not much point modeling the details of errors that we can prove are never emitted, but yes, we could  in principle get to a fully detailed model in time; the maybe-temporary existence of `FakeError` or `ImpossibleError` will enable us to get there incrementally.  Regarding postulating `fakeErr : ErrLog`, I thought that was how I started and you didn't like it :smile:.  Maybe I'm confused though.  In any case, it seems reasonable to me, and this conversation has made me want to make our type signatures look identical again :smile:.

**Comment by mark-moir** [comment]
@cwjnkins AFAIAC, this is ready to merge.  Merge at will!  Thanks.

**Pull request merged**
